keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,2333 @@
+          #- name: high     # class "high" is not allowed at all
+    container:
+      - name: qos-foo
+        classes:
+          - name: low-prio
+          - name: normal-prioa
+          #- name: high-prio    # high-prio is not allowed at all
+      - name: vendor.example/xyz-qos
+        classes:
+          - name: low-prio
+          - name: normal
+          - name: high-prio
+            capacity: 1     # high-prio is allowed but with very limited usage
+```
+
+#### Set ulimits
+
+As a cluster administrator I want to enable per-application control of system
+resource usage /such as maximum number of processes or open files) exposed as
+ulimits/rlimits in a Linux/UNIX system. However, I don't want to give the users
+full control of the exact numeric parameters but specify pre-defined classes of
+limits for different workload profiles.
+
+Currently, ulimits can be set via an
+[NRI plugin](https://github.com/containerd/nri/tree/main/plugins/ulimit-adjuster)
+but that relies on Pod annotations. Using QoS-class resources would give better
+user interface and control to the cluster administrator.
+
+#### Possible future scenarios
+
+This section speculates on possible future uses of the QoS-class resources
+mechanism.
+
+##### Pass on Pod QoS class to the runtime
+
+The [Pod QoS class][pod-qos-class] could be communicated to the container
+runtime as a special Kubernetes-specific QoS-class resource.
+
+Information about Pod QoS class is currently internal to kubelet and not
+visible to the container runtime. However, container runtimes (CRI-O, at least)
+are already depending on this information and currently determining it
+indirectly by evaluating other CRI parameters.
+
+This change would make the information about Pod QoS explicit and would allow
+elimination of unreliable code paths in runtimes, for example.
+
+Pod QoS class would not be advertised as one of the available QoS-class
+resources in NodeStatus. Also, users would not be allowed to request it in the
+PodSpec which would be enforced by admission checks in the api-server and
+kubelet. In other words, this would be purely informational, aimed for the
+container runtime.
+
+```diff
++ // Kubernetes-managed QoS resources. These are only informational to the runtime
++ // and no CRI requests should fail because of these.
++const (
++       // QOSResourcePodQOS is the Kubernetes Pod Quality of Service Class.
++       // Possible values are "Guaranteed", "Burstable" and "BestEffort".
++       QOSResourcePodQOS = "pod-qos"
++)
+```
+
+##### Kubernetes-managed QoS-class resources
+
+It would be possible to have QoS-class resources that would be managed by
+Kubernetes/kubelet instead of the container runtime. If we specify and manage
+official well-known QoS-class resource names in the API it would be possible to
+specify Kubernetes-internal names that the container runtime would know to
+ignore (or not try to manage itself).
```

Comment:
Just a note that this section is speculation for possible future usages, not fully designed by any means, and not proposed to be implemented.

I suppose it would be kubelet who's responsible for keeping the node status. There are multiple options how to distinguish them but I think possibly the most obvious is to have a well-known prefix (like `k8s.io/`) to have for K8s-managed QoS-resources. Other options include e.g. a) Just list the K8s-managed QoS in the API (consts) but then there's a version skew problem (if smth is added to the API, older runtimes are not aware of this) b) Kubelet just filters out these from the CRI request (i.e. the runtime would be totally unaware of these) but there are probably cases where the runtime would like to know about these (as informational) c) Kubelet tells the CRI runtime what are the K8s-managed QoS-resources (like the UpdateRuntimeConfig rpc) but this feels complex

I added a mention about this in this section as well as the "Consts" section below.

Reply:
@swatisehgal after a second thought I changed the proposed naming so that unprefixed names (like `podqos` would be reserved for Kubernetes-managed "native" QoS resources and `k8s.io/` prefix would be reserved for official/canonical runtime-managed QoS resources.

Reply:
This makes sense, Thanks for the clarification.

Reply:
> @swatisehgal after a second thought I changed the proposed naming so that unprefixed names (like `podqos` would be reserved for Kubernetes-managed "native" QoS resources and `k8s.io/` prefix would be reserved for official/canonical runtime-managed QoS resources.

That's the opposite of how we usually do it; for labels, `example-thing` and `foo.bar.23` are private use for a cluster operator to pick their own names, whereas `kubernetes.io/example` and `qos.kubernetes.io/example-2` are official.

Reply:
We could do that if using an unrecognized name with no prefix is an API error - is that the plan?

Reply:
@marquiz to confirm, but in my opinion: yes. un-preffixed we should have clean list constants that api server and kubelet would verify ownership. for reserved prefix we might want to agree for kubelet/runtime validation schema if they are ok to share k8s.io/kubernetes.io prefix. for anything else, some simple validation for naming schema like API server does for extended resources names.